addon/components/ember-collection.js

```diff
@@ -80,17 +81,24 @@ export default Ember.Component.extend({
 
   updateItems(){
     this._cellLayout = this.getAttr('cell-layout');
-    var items = this.getAttr('items');
-
-    if (this._items !== items) {
-      if (this._items && this._items.removeObserver) {
-        this._items.removeObserver('[]', this, this._needsRevalidate);
+    var rawItems = this.getAttr('items');
+
+    if (this._rawItems !== rawItems) {
+      if (this._items && this._items.removeArrayObserver) {
+        this._items.removeEnumerableObserver(this, {
+          willChange: Ember.K,
+          didChange: '_needsRevalidate'
+        });
       }
-
+      this._rawItems = rawItems;
+      var items = Ember.A(rawItems);
       this.set('_items', items);
 
-      if (items && items.addObserver) {
-        items.addObserver('[]', this, this._needsRevalidate);
+      if (items && items.addEnumerableObserver) {
```

Comment:
That sounds like a bug in ArrayProxy to be honest, also, addEnumerableObserver is soft deprecated and marked as private in 2.0.  We can use addArrayObserver but to be honest I was adding a indirection here called dataSource so people could roll their own sparse array as part of the layout bin refactor but I'd like to pull this in for the mean time.